src/scenarios/emqttb_scenario_persistent_session.erl

```diff
@@ -0,0 +1,279 @@
+%%
+%% Licensed under the Apache License, Version 2.0 (the "License");
+%% you may not use this file except in compliance with the License.
+%% You may obtain a copy of the License at
+%%
+%%     http://www.apache.org/licenses/LICENSE-2.0
+%%
+%% Unless required by applicable law or agreed to in writing, software
+%% distributed under the License is distributed on an "AS IS" BASIS,
+%% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+%% See the License for the specific language governing permissions and
+%% limitations under the License.
+%%--------------------------------------------------------------------
+-module(emqttb_scenario_persistent_session).
+
+-behavior(emqttb_scenario).
+
+%% The scenario alternates between `subscribe_phase' and
+%% `publish_phase'
+%%
+%% Publish stage ends after a set time interval
+%%
+%% Subscribe stage after the number of consumed messages per second
+%% drops to 0
+%%
+%% KPIs:
+%% - number of messages published during publish phase
+%% - time to consume consume the messages
+%% - subscription rate
+
+%% behavior callbacks:
+-export([ name/0
+        , model/0
+        , run/0
+        ]).
+
+%% internal exports:
+-export([]).
+
+-export_type([]).
+
+-include("emqttb.hrl").
+-include("../framework/emqttb_internal.hrl").
+-include_lib("typerefl/include/types.hrl").
+
+-import(emqttb_scenario, [complete/1, loiter/0, my_conf/1, my_conf_key/1, set_stage/2, set_stage/1]).
+
+%%================================================================================
+%% Type declarations
+%%================================================================================
+
+-define(PUB_GROUP, pers_sess_pub_group_pub).
+-define(SUB_GROUP, pers_sess_pub_group_sub).
+
+-define(PUB_THROUGHPUT, emqttb_pers_sess_pub_throughput).
+-define(SUB_THROUGHPUT, emqttb_pers_sess_sub_throughput).
+
+-record(s,
+        { produced = 0
+        , consumed = 0
+        , to_consume = 0
+        }).
+
+%%================================================================================
+%% behavior callbacks
+%%================================================================================
+
+name() ->
+  persistent_session.
+
+model() ->
+  #{ pub =>
+       #{ qos =>
+            {[value, cli_param],
+             #{ oneliner => "QoS of the published messages"
+              , type => emqttb:qos()
+              , default => 1
+              , cli_operand => "pub-qos"
+              }}
+        , msg_size =>
+            {[value, cli_param],
+             #{ oneliner => "Size of the published message in bytes"
+              , type => non_neg_integer()
+              , cli_operand => "size"
+              , cli_short => $s
+              , default => 256
+              }}
+        , pubinterval =>
+            {[value, cli_param],
+             #{ oneliner => "Message publishing interval"
+              , type => emqttb:interval()
+              , default_ref => [interval]
```

Comment:
Is it possible to know which `interval` this refers to?  Is it the interval in the root model?

Also, is `default_ref` to use a default value set in another mnode?

Reply:
Yes, default_ref takes the default value from another value. Here `[interval]` is the global key.